src/serac/numerics/functional/shape_aware_functional.hpp

```diff
@@ -0,0 +1,485 @@
+/// @endcond
+
+/**
+ * @brief This is a small wrapper around serac::Functional for shape-displaced domains of integration
+ *
+ * If a finite element kernel is defined on a domain \f$x = X + p\f$ where \f$X\f$ is the reference
+ * configuration and \f$p\f$ is a shape displacement field. This shape displacement is typically a input
+ * parameter used in shape optimization problems. This wrapper correctly modifies domain and boundary
+ * integrals for L2 and H1 trial functions and H1, L2, and QOI (double) test functions for vector-valued
+ * H1 shape displacement fields.
+ *
+ * See serac::Functional for more details about the underlying finite element abstraction and serac::HeatTransfer
+ * for an example of how to use this wrapper.
+ *
+ * @tparam test The space of test function to use
+ * @tparam shape The space of the shape displacement function to use
+ * @tparam trials The space of trial functions to use
+ * @tparam exec whether to carry out calculations on CPU or GPU
+ */
+template <typename test, typename shape, typename... trials, ExecutionSpace exec>
+class ShapeAwareFunctional<shape, test(trials...), exec> {
+  /// @brief The compile-time trial function finite element spaces
+  static constexpr tuple<trials...> trial_spaces{};
+
+  /// @brief The compile-time test function finite element space
+  static constexpr test test_space{};
+
+  /// @brief The compile-time shape displacement finite element space
+  static constexpr shape shape_space{};
+
+  /// @brief The number of input trial functions
+  static constexpr uint32_t num_trial_spaces = sizeof...(trials);
+
+public:
+  /**
+   * @brief Constructs using @p mfem::ParFiniteElementSpace objects corresponding to the test/trial spaces
+   *
+   * @tparam test_type The finite element space of the test functions (used to disable this method if called with test =
+   * double)
+   *
+   * @param[in] shape_fes The shape displacement finite element space
+   * @param[in] test_fes The (non-qoi) test finite element space
+   * @param[in] trial_fes The trial finite element spaces
+   */
+  template <typename test_type = test, typename = std::enable_if_t<!std::is_same_v<double, test_type>>>
+  ShapeAwareFunctional(const mfem::ParFiniteElementSpace* shape_fes, const mfem::ParFiniteElementSpace* test_fes,
+                       std::array<const mfem::ParFiniteElementSpace*, num_trial_spaces> trial_fes)
+  {
+    static_assert(test_space.family != Family::HDIV, "Shape-aware functional not implemented for HDiv test functions");
+    static_assert(test_space.family != Family::HCURL,
+                  "Shape-aware functional not implemented for HCurl test functions");
+    static_assert(shape_space.family == Family::H1, "Only H1 spaces allowed for shape displacements");
+
+    for_constexpr<num_trial_spaces>([](auto i) {
+      static_assert(get<i>(trial_spaces).family != Family::HDIV,
+                    "Shape-aware functional not implemented for HDiv trial functions");
```

Comment:
Since no part of serac currently supports Hdiv, I think there are more appropriate places to put this kind of static assertion (e.g. either `serac::Functional` and/or `serac::FiniteElementVector`)